test/compilable/b12647.d

```diff
@@ -0,0 +1,9 @@
+void assertThrown12647(E)(lazy E expression)
+{
+    expression();
+}
+
+void test12647() nothrow pure @safe @nogc
+{
+    assertThrown12647(0);
+}
```

Comment:
Not quite, I'll provide examples.

First, there are several problems at work here. I'll start with just functions (without templates), because they are easier to reason about.
This currently compiles:
```D
void test2(lazy Exception e) nothrow pure @safe @nogc {}
void test() nothrow pure @safe @nogc { test2(new Exception(null)); }
```

Because `lazy` does not have any annotation (by annotations, I mean `nothrow`, `pure` and the likes), but the delegate doesn't get called.

However, the following does not:
```D
void test2(lazy Exception e) nothrow pure @safe @nogc { auto exc = e; }
void test() nothrow pure @safe @nogc { test2(new Exception(null)); }
```

For the very same reason (no annotation). The obvious fix is to require the type of the `lazy`'s delegate to match the function's annotations. But that would break the first case. And since the first case is valid code, I'm afraid we have to go with deprecations once again.